packages/app/src/cli/utilities/developer-platform-client/partners-client.test.ts

```diff
@@ -175,6 +185,81 @@ describe('fetchApp', async () => {
 
     // Then
     await expect(got).rejects.toThrowError(new NoOrgError(testPartnersUserSession.accountInfo))
-    expect(partnersRequest).toHaveBeenCalledWith(FindOrganizationQuery, 'token', {id: ORG1.id}, undefined, undefined)
+    expect(partnersRequest).toHaveBeenCalledWith(FindOrganizationQuery, 'token', {id: ORG1.id}, undefined, undefined, {
+      type: 'token_refresh',
+      handler: expect.any(Function),
+    })
+  })
+})
+
+describe('Token refresh', () => {
+  let partnersClient: PartnersClient
+  let sessionMock: PartnersSession
+  const partnersRequestMock = vi.mocked(partnersRequest)
+  const query = `query { doesnt { matter } }`
+  const variables = undefined
+  const response = {}
+  const newToken = 'new-token'
+  let refreshedToken: string | undefined
+
+  beforeEach(() => {
+    sessionMock = testPartnersUserSession
+    partnersClient = new PartnersClient(sessionMock)
+    vi.spyOn(partnersClient, 'refreshToken').mockImplementation(vi.fn())
+    ;(partnersClient.refreshToken as Mock).mockResolvedValue(newToken)
+
+    partnersRequestMock.mockImplementation(
+      async (_query, _token, _variables, _headers, _preferredBehaviour, handler) => {
+        if (handler?.type === 'token_refresh') {
+          refreshedToken = (await handler.handler()).token
+        }
+        return Promise.resolve(response)
+      },
+    )
+  })
+
+  test('refreshes token once if the handler is called', async () => {
+    await expect(partnersClient.request(query, variables)).resolves.toEqual(response)
+
+    expect(refreshedToken).toBe(newToken)
+    expect(partnersClient.refreshToken).toHaveBeenCalledOnce()
+    expect(partnersRequestMock).toHaveBeenCalledTimes(1)
+  })
+
+  test('if token refresh is called twice, the token returns undefined', async () => {
+    partnersRequestMock.mockImplementation(
+      async (_query, _token, _variables, _headers, _preferredBehaviour, handler) => {
+        if (handler?.type === 'token_refresh') {
+          // Call handler twice, simulating a failure of the first call.
+          await handler.handler()
+          refreshedToken = (await handler.handler()).token
+        }
+        return Promise.resolve(response)
+      },
+    )
+
+    await expect(partnersClient.request(query, variables)).resolves.toEqual(response)
+
+    expect(refreshedToken).toBe(undefined)
+    expect(partnersClient.refreshToken).toHaveBeenCalledOnce()
+    expect(partnersRequestMock).toHaveBeenCalledTimes(1)
+  })
+
+  test('if multiple calls to partnerClient are made while token refresh is happening, the second call fails', async () => {
+    let refreshedToken: string | undefined
+    ;(partnersClient.refreshToken as Mock).mockImplementationOnce(async () => {
+      // Just sleep so the second call immediately trips the race condition guard clause
+      await new Promise((resolve) => setTimeout(resolve, 3000))
```

Comment:
In terms of correctness of sleeping, answer is: it depends. Actually sleeping for a while -- i.e ending up with a slow test -- is not ideal. We should only reach for that if truly needed.

Starting to sleep but not actually finishing the sleep, or doing a sleep that's pretty short -- generally no problem. Ultimately the test speed is what matters and these both round down to fast. If doing that to try to fudge some non-deterministic behaviour, that's not good.

If you need to test something that involves a noticeable passage of time, vitest's timer mocking is best: https://vitest.dev/guide/mocking.html#timers 

I think in this case, providing my comment below makes sense, it looks like a case for timer mocking.